File: 백준/Gold/1967. 트리의 지름/트리의 지름.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

setrecursionlimit(10**9)
input = stdin.readline


def dfs(now,w):
    global graph
    
    for (nxt,cost) in graph[now]:
        
        if dist[nxt] == -1:
            dist[nxt] = w + cost
            dfs(nxt,dist[nxt])
    
    
    

n = int(input())
graph =[[] for _ in range(n+1)]

for _ in range(n-1):
    p,c,w = map(int,input().split())
    graph[p].append((c,w))
    graph[c].append((p,w))


dist = [-1 for _ in range(n+1)]
dist[1] = 1

dfs(1,0) # 루트에서 dfs  진행 후, 가장 먼 지점 찾아 낸 후

start = dist.index(max(dist))
dist = [-1 for _ in range(n+1)]

# 그 지점을 기준으로 다시 dfs 돌려 지름을 구한다
dist[start] = 0
dfs(start,0) 

print(max(dist))




    

File: 백준/Gold/1967. 트리의 지름/트리의 지름.swift
let n = Int(readLine()!)!

struct Node {
    let dest: Int
    let cost: Int
    
    init(_ dest:Int,_ cost:Int) {
        self.dest = dest
        self.cost = cost
    }
}

var graph: [[Node]] = Array(repeating: [], count: n+1)

for _ in 0..<n-1 {
    let input = readLine()!.split{$0 == " "}.map{Int($0)!}
    
    graph[input[0]].append(Node(input[1],input[2]))
    graph[input[1]].append(Node(input[0],input[2]))
    
}

var dist : [Int] = [Int](repeating: -1, count: n+1)

func dfs(_ now: Int, _ cost:Int) {
    
    for node in graph[now] {
        
        if dist[node.dest] == -1 {
            dist[node.dest] = cost + node.cost
            dfs(node.dest, dist[node.dest])
        }
        
    }
    
}

dist[1] = 0
dfs(1,0)

let dot1: Int = dist.distance(from: 0, to: dist.firstIndex(of: dist.max()!)!)

dist = [Int](repeating: -1, count: n+1)
dist[dot1] = 0
dfs(dot1, 0)

print(dist.max()!)
